feat(utils): support multiple certificates in resolveServerUrls

monam2 opened the pull request:
### Description

Fixes #20681

Added support for multiple certificates in the resolveServerUrls function. Previously, only single certificates were supported, limiting HTTPS server configuration options.

- Uses arraify to process certificate arrays and handle multiple certificates properly.
- Added bufferify function for safe certificate conversion to Buffer format.
- Implemented try-catch blocks to gracefully handle invalid certificates and prevent crashes.

All tests pass (unit, integration, and build tests all pass locally)

---

## Test Coverage Added

- No Certificate
- Single Certificate  
- Multiple Certificates
- IPv6 Support
- Mixed Configurations (IPv4/IPv6 address families)
- Invalid Certificate Handling

---

**Note**: Some test parameters use `any` types due to complex interface typing challenges. Type improvements are welcome.

monam2 committed:
feat(utils): support multiple certificates in resolveServerUrls

monam2 committed:
test: update multiple certs handling tests for IPv4 and IPv6

monam2 commented:
Sorry for the confusion in the test cases.
I've updated the multiple certs tests to use the same cert twice and removed the mixed IPv4/IPv6 test case since it's not directly related to this work.

Any feedback on the certificate or implementation setup would be welcome.

sapphi-red commented on packages/vite/src/node/utils.ts:
```diff
@@ -1045,19 +1049,27 @@ export function resolveServerUrls(
+  const hostnameFromCert = certs
+    .map((cert) => {
+      try {
+        const buf = bufferify(cert)
+        return new crypto.X509Certificate(buf)
```
```suggestion
        return new crypto.X509Certificate(cert)
```
I think this conversion is handled internally by `X509Certificate`.

monam2 replied:
I added this conversion mainly because of the mock work during testing, but if we extract it into a helper function as suggested, I think it can be removed as you said. Thanks!

sapphi-red commented on packages/vite/src/node/utils.ts:
```diff
@@ -1045,19 +1049,27 @@ export function resolveServerUrls(
+      }
+    })
+    .flatMap((cert) =>
+      cert && cert.subjectAltName
```
```suggestion
      cert?.subjectAltName
```

sapphi-red commented on packages/vite/src/node/utils.ts:
```diff
@@ -1045,19 +1049,27 @@ export function resolveServerUrls(
-  const hostnameFromCert = cert?.subjectAltName
-    ? extractHostnamesFromSubjectAltName(cert.subjectAltName)
-    : []
+  const certs = httpsOptions?.cert ? arraify(httpsOptions.cert) : []
+  const hostnameFromCert = certs
+    .map((cert) => {
+      try {
+        const buf = bufferify(cert)
+        return new crypto.X509Certificate(buf)
+      } catch {
+        return null
+      }
+    })
+    .flatMap((cert) =>
+      cert && cert.subjectAltName
+        ? extractHostnamesFromSubjectAltName(cert.subjectAltName)
+        : [],
+    )
```
Would you extract this part as a function so that we don't have to use mocks in the tests?

sapphi-red reviewed (changes requested):
Thanks!

monam2 committed:
feat(utils): extract hostnames from certs via helper and use in resolveServerUrls

monam2 commented:
@sapphi-red Thanks for the suggestion — I’ve updated the code accordingly. I extracted the logic into a dedicated helper function and also renamed a few variables for clarity. PTAL.

monam2 committed:
test(utils): cover extractHostnamesFromCerts with single and multiple certs

bluwy commented on packages/vite/src/node/__tests__/utils.spec.ts:
```diff
@@ -908,3 +910,204 @@ describe('getServerUrlByHost', () => {
+
+  test('should extract hostnames from multiple certificates', () => {
+    const httpsOptions = { cert: [createWorkingCert, createWorkingCert] } as any
+    const result = extractHostnamesFromCerts(httpsOptions.cert)
+
+    expect(result).toStrictEqual([
+      'localhost',
+      'foo.localhost',
+      'vite.vite.localhost',
+      'localhost',
+      'foo.localhost',
+      'vite.vite.localhost',
+    ])
```
Should we make `extractHostnamesFromCerts` only return unique hosts? IIUC the duplicated ones would end up showing multiple times when printed otherwise.

monam2 committed:
fix(utils): dedupe hostnames returned by extractHostnamesFromCerts

monam2 committed:
test(utils): hoist WORKING_TEST_CERT and update cert-related tests

monam2 commented on packages/vite/src/node/__tests__/utils.spec.ts:
```diff
@@ -908,3 +910,201 @@ describe('getServerUrlByHost', () => {
+WEIBLwjKCAscrtyayXWp6zUTmgvb8PQf//3Mh2DiEngAi3WI/nL+8Y0RkwbvxBar
+X2JN
+-----END CERTIFICATE-----
+    `.trim() as any
```
@bluwy I used the test certificate added in #19317. I uploaded it to the top constant 'WORKING_TEST_CERT' and updated the related tests for reuse in several cases. PTAL

bluwy replied:
Ah I didn't notice that the cert was generated from a previous PR. Thanks for looking into it.

bluwy commented on packages/vite/src/node/utils.ts:
```diff
@@ -993,6 +993,33 @@ export async function resolveHostname(
   return { host, name }
 }
 
+export function bufferify(buffer: string | Buffer): Buffer {
+  return Buffer.isBuffer(buffer) ? buffer : Buffer.from(buffer)
+}
```
Sorry for the back and forth, but it seems like this function is unused now.

monam2 replied:
I removed that function

monam2 committed:
refactor(utils): remove unused bufferify function

sapphi-red committed:
chore: update

EmperorArthur reviewed (commented):
I really appreciate you taking the time to work on this.

Here are a few small things that you might want to consider tweaking.  It's up to you and the maintainers of course, but thought I could try to contribute a little bit.

sapphi-red commented on packages/vite/src/node/utils.ts:
```diff
@@ -993,6 +993,29 @@ export async function resolveHostname(
+      }
+    })
+    .flatMap((cert) =>
+      cert?.subjectAltName
```
TypeScript cannot infer the types with
```ts
    .filter((cert) => Boolean(cert?.subjectAltName))
    .flatMap((cert) => extractHostnamesFromSubjectAltName(cert.subjectAltName)) // cert may be null
```
so I think it's better to keep it as-is.
